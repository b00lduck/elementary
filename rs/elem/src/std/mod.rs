mod primitives;

pub mod prelude {
    pub use super::primitives::*;
}
