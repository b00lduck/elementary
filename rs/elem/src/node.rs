use serde::{Deserialize, Serialize};
use std::hash::{DefaultHasher, Hash, Hasher};

#[derive(Serialize, Deserialize)]
pub struct NodeRepr {
    pub hash: i32,
    pub kind: String,
    pub props: serde_json::Map<String, serde_json::Value>,
    pub output_channel: u32,
    pub children: Vec<NodeRepr>,
}

pub fn create_node(
    kind: &str,
    props: serde_json::Map<String, serde_json::Value>,
    children: Vec<NodeRepr>,
) -> NodeRepr {
    let mut hasher = DefaultHasher::new();

    kind.hash(&mut hasher);
    props.hash(&mut hasher);

    for child in children.iter() {
        child.hash.hash(&mut hasher);
    }

    NodeRepr {
        hash: hasher.finish() as i32,
        kind: kind.to_string(),
        props,
        output_channel: 0,
        children,
    }
}

pub struct ShallowNodeRepr {
    hash: i32,
    kind: String,
    props: serde_json::Map<String, serde_json::Value>,
    output_channel: u32,
    children: Vec<i32>,
}

pub fn shallow_clone(node: &NodeRepr) -> ShallowNodeRepr {
    ShallowNodeRepr {
        hash: node.hash,
        kind: node.kind.clone(),
        props: node.props.clone(),
        output_channel: node.output_channel,
        children: node.children.iter().map(|n| n.hash).collect::<Vec<i32>>(),
    }
}
