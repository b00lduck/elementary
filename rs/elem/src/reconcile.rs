use std::cmp::Ordering;
use std::collections::{BTreeMap, HashSet, VecDeque};

use crate::node::{NodeRepr, ShallowNodeRepr};
use serde::ser::{Serialize, SerializeTuple, Serializer};
// use rpds::RedBlackTreeMap;

#[derive(Debug, Eq, PartialEq)]
pub enum Instruction {
    Create(i32, String),
    AppendChild(i32, i32, u32),
    SetProperty(i32, String, serde_json::Value),
    ActivateRoots(Vec<i32>),
    Commit,
}

impl PartialOrd for Instruction {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Instruction {
    fn cmp(&self, other: &Self) -> Ordering {
        match (self, other) {
            (Instruction::Create(_, _), Instruction::Create(_, _)) => Ordering::Equal,
            (Instruction::Create(_, _), _) => Ordering::Less,
            (_, _) => Ordering::Equal,
        }
    }
}

impl Serialize for Instruction {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        match self {
            Self::Create(hash, kind) => {
                let mut tup = serializer.serialize_tuple(3)?;
                let tag: i32 = 0;
                tup.serialize_element(&tag)?;
                tup.serialize_element(hash)?;
                tup.serialize_element(kind)?;
                tup.end()
            }
            Self::AppendChild(parent_hash, child_hash, output_channel) => {
                let mut tup = serializer.serialize_tuple(4)?;
                let tag: i32 = 2;
                tup.serialize_element(&tag)?;
                tup.serialize_element(parent_hash)?;
                tup.serialize_element(child_hash)?;
                tup.serialize_element(output_channel)?;
                tup.end()
            }
            Self::SetProperty(hash, key, value) => {
                let mut tup = serializer.serialize_tuple(4)?;
                let tag: i32 = 3;
                tup.serialize_element(&tag)?;
                tup.serialize_element(hash)?;
                tup.serialize_element(key)?;
                tup.serialize_element(value)?;
                tup.end()
            }
            Self::ActivateRoots(roots) => {
                let mut tup = serializer.serialize_tuple(2)?;
                let tag: i32 = 4;
                tup.serialize_element(&tag)?;
                tup.serialize_element(roots)?;
                tup.end()
            }
            Self::Commit => {
                let mut tup = serializer.serialize_tuple(1)?;
                let tag: i32 = 5;
                tup.serialize_element(&tag)?;
                tup.end()
            }
        }
    }
}

pub fn reconcile(
    node_map: &mut BTreeMap<i32, ShallowNodeRepr>,
    roots: &Vec<NodeRepr>,
) -> Vec<Instruction> {
    let mut visited: HashSet<i32> = HashSet::new();
    let mut queue: VecDeque<&NodeRepr> = VecDeque::from_iter(roots.iter());
    let mut instructions: Vec<Instruction> = Vec::new();

    loop {
        match queue.pop_front() {
            None => break,
            Some(next) => {
                if visited.contains(&next.hash) {
                    continue;
                }

                // Mount
                let existing = node_map.entry(next.hash).or_insert_with(|| {
                    // Create node
                    instructions.push(Instruction::Create(next.hash, next.kind.clone()));

                    // Append child
                    for child in next.children.iter() {
                        instructions.push(Instruction::AppendChild(
                            next.hash,
                            child.hash,
                            child.output_channel,
                        ));
                    }

                    next.into()
                });

                // Props
                for (name, value) in &next.props {
                    let do_write = match existing.props.get(name) {
                        Some(v) => v != value,
                        None => true,
                    };

                    if do_write {
                        existing
                            .props
                            .entry(name)
                            .and_modify(|e| *e = value.clone())
                            .or_insert(value.clone());

                        instructions.push(Instruction::SetProperty(
                            next.hash,
                            name.clone(),
                            value.clone(),
                        ));
                    }
                }

                // Visit children
                for child in next.children.iter() {
                    queue.push_back(child);
                }

                visited.insert(next.hash);
            }
        }
    }

    // Activate roots
    instructions.push(Instruction::ActivateRoots(
        roots.iter().map(|n| n.hash).collect::<Vec<i32>>(),
    ));

    // Commit
    instructions.push(Instruction::Commit);

    // Sort so that creates land before appends, etc
    instructions.sort();
    instructions
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::std::prelude::*;

    #[test]
    fn basic() {
        let mut node_map = BTreeMap::new();
        let graph = vec![root(phasor(constant!({key: None, value: 110.0})))];
        let instructions = reconcile(&mut node_map, &graph);

        insta::assert_json_snapshot!(instructions);
    }

    #[test]
    fn distinguish_by_props() {
        let voice =
            |path| sample!({key: None, path: path}, train(constant!({key: None, value: 1.0})));
        let mut node_map = BTreeMap::new();
        let graph = vec![
            voice(String::from("test.wav")),
            voice(String::from("test2.wav")),
        ];
        let instructions = reconcile(&mut node_map, &graph);

        insta::assert_json_snapshot!(instructions);
    }

    #[test]
    fn structural_equality_with_value_change() {
        let mut node_map = BTreeMap::new();

        let voice = |key, freq| {
            sin(mul2(
                cv(2.0 * std::f64::consts::PI),
                phasor(constant!({key: Some(key), value: freq})),
            ))
        };

        let graph = vec![add2(
            voice(String::from("v1"), 110.0),
            voice(String::from("v2"), 111.0),
        )];

        // For the sake of this test, we don't care about this first pass
        let _ = reconcile(&mut node_map, &graph);

        // Now we build a second graph aiming to see that the instruction
        // set produced contains only a single SetProperty
        let graph2 = vec![add2(
            voice(String::from("v1"), 112.0),
            voice(String::from("v2"), 111.0),
        )];

        let instructions = reconcile(&mut node_map, &graph2);
        insta::assert_json_snapshot!(instructions);
    }
}
