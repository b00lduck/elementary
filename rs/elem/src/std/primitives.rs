use crate::node::{create_node, NodeRepr};
use serde::{Deserialize, Serialize};
use serde_json::json;

pub fn root(x: NodeRepr) -> NodeRepr {
    create_node(
        "root",
        json!({"channel": 0.0}).as_object().unwrap().clone(),
        vec![x],
    )
}

pub fn sin(x: NodeRepr) -> NodeRepr {
    create_node("sin", Default::default(), vec![x])
}

pub fn mul2(x: NodeRepr, y: NodeRepr) -> NodeRepr {
    create_node("mul", Default::default(), vec![x, y])
}

pub fn add2(x: NodeRepr, y: NodeRepr) -> NodeRepr {
    create_node("add", Default::default(), vec![x, y])
}

pub fn phasor(rate: NodeRepr) -> NodeRepr {
    create_node("phasor", Default::default(), vec![rate])
}

#[derive(Serialize, Deserialize)]
pub struct ConstNodeProps {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub key: Option<String>,
    pub value: f64,
}

pub fn constant(props: &ConstNodeProps) -> NodeRepr {
    create_node(
        "const",
        serde_json::to_value(&props)
            .unwrap()
            .as_object()
            .unwrap()
            .clone(),
        vec![],
    )
}

#[macro_export]
macro_rules! constant {
    // Match the macro pattern with a key-value pair in the first argument
    ({$($key:ident: $value:expr),*}) => {
        {
            // Create the props struct with the provided key-value pairs
            let props = ConstNodeProps { $($key: $value),* };

            // Call the constant function with the constructed props
            constant(&props)
        }
    };
}

pub use crate::constant;

pub fn cv(x: f64) -> NodeRepr {
    constant!({key: None, value: x})
}

pub fn le(x: NodeRepr, y: NodeRepr) -> NodeRepr {
    create_node("le", Default::default(), vec![x, y])
}

pub fn train(x: NodeRepr) -> NodeRepr {
    le(phasor(x), cv(0.5))
}

#[derive(Serialize, Deserialize)]
pub struct SampleNodeProps {
    pub key: Option<String>,
    pub path: String,
}

pub fn sample(props: &SampleNodeProps, gate: NodeRepr) -> NodeRepr {
    create_node(
        "sample",
        serde_json::to_value(&props)
            .unwrap()
            .as_object()
            .unwrap()
            .clone(),
        vec![gate],
    )
}

#[macro_export]
macro_rules! sample {
    // Match the macro pattern with a key-value pair in the first argument
    ({$($key:ident: $value:expr),*}, $gate:expr) => {
        {
            // Create the props struct with the provided key-value pairs
            let props = SampleNodeProps { $($key: $value),* };

            // Call the constant function with the constructed props
            sample(&props, $gate)
        }
    };
}

pub use crate::sample;
