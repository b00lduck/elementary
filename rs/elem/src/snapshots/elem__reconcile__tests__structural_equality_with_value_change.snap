---
source: elem/src/reconcile.rs
expression: instructions
---
[
  [
    3,
    531638123,
    "value",
    112.0
  ],
  [
    4,
    [
      -210358393
    ]
  ],
  [
    5
  ]
]
